includes/utils/class-amp-carousel.php

```diff
@@ -0,0 +1,160 @@
+<?php
+/**
+ * Class AMP_Carousel.
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Carousel
+ *
+ * Gets the markup for an <amp-carousel>.
+ *
+ * @since 1.4.1
+ */
+class AMP_Carousel {
+
+	/**
+	 * Value used for width of amp-carousel.
+	 *
+	 * @var int
+	 */
+	const FALLBACK_WIDTH = 600;
+
+	/**
+	 * Value used for height of amp-carousel.
+	 *
+	 * @var int
+	 */
+	const FALLBACK_HEIGHT = 480;
+
+	/**
+	 * An object representation of the DOM.
+	 *
+	 * @var DOMDocument
+	 */
+	protected $dom;
+
+	/**
+	 * Instantiates the class.
+	 *
+	 * @param DOMDocument $dom The dom to use to create a carousel.
+	 */
+	public function __construct( $dom ) {
+		$this->dom = $dom;
+	}
+
+	/**
+	 * Creates and gets an <amp-carousel> with the given images and captions.
+	 *
+	 * @param array[] {
+	 *     The images and their respective captions, if any.
+	 *
+	 *     @type DOMElement  $image A representation of an image, in index 0.
+	 *     @type string|null $caption The caption for that image, if any, in index 1.
+	 * }
+	 * @return DOMElement A representation of the <amp-carousel>.
+	 */
+	public function create_and_get( $images_and_captions ) {
```

Comment:
It'd be better if `$images_and_captions` were an associative array of images to their respective captions.

But a `DOMElement` (the image) isn't allowed as key of an associative array.

Reply:
That's pretty slick. I suppose my only hesitation is overly engineering an API for this which we don't necessarily want others to use externally, until we've given it it enough thought into how it relates to the other pieces. If we just use a simple array as the input, then it is more flexible in the case we decide to go a different direction. But then again, if we don't want people to be using this class yet anyway, then I don't suppose it matters. Just make sure it is marked with `@internal`.

If we do go this route, then it should definitely fall under the `AMP` namespace and the autoloader should automatically locate the filename purely based off of the name of the class/interface/trait. If we start creating a ton of little PHP files, we don't want to be having to embiggen the `\AMP_Autoloader::$classmap` endlessly.

Reply:
Yes, I agree with marking as `@internal` for now. We should slowly build up our understanding and infrastructure with regards to an OOP approach of the plugin, and only really advertise an explicit set of interfaces with v2.0.